python/ray/serve/_private/deployment_scheduler.py

```diff
@@ -0,0 +1,266 @@
+        else:
+            self._gcs_client = GcsClient(address=ray.get_runtime_context().gcs_address)
+
+    def on_deployment_created(
+        self,
+        deployment_name: str,
+        scheduling_policy: Union[
+            SpreadDeploymentSchedulingPolicy, DriverDeploymentSchedulingPolicy
+        ],
+    ):
+        """This is called whenver a new deployment is created."""
+        assert deployment_name not in self._pending_replicas
+        assert deployment_name not in self._launching_replicas
+        assert deployment_name not in self._recovering_replicas
+        assert deployment_name not in self._running_replicas
+        self._deployments[deployment_name] = scheduling_policy
+
+    def on_deployment_deleted(self, deployment_name: str):
+        """This is called whenver a deployment is deleted."""
+        assert not self._pending_replicas[deployment_name]
+        del self._pending_replicas[deployment_name]
+
+        assert not self._launching_replicas[deployment_name]
+        del self._launching_replicas[deployment_name]
+
+        assert not self._recovering_replicas[deployment_name]
+        del self._recovering_replicas[deployment_name]
+
+        assert not self._running_replicas[deployment_name]
+        del self._running_replicas[deployment_name]
+
+        del self._deployments[deployment_name]
+
+    def on_replica_stopping(self, deployment_name: str, replica_name: str):
+        """This is called whenver a deployment replica is being stopped."""
+        self._pending_replicas[deployment_name].pop(replica_name, None)
+        self._launching_replicas[deployment_name].pop(replica_name, None)
+        self._recovering_replicas[deployment_name].discard(replica_name)
+        self._running_replicas[deployment_name].pop(replica_name, None)
+
+    def on_replica_running(self, deployment_name: str, replica_name: str, node_id: str):
+        """This is called whenver a deployment replica is running with known node id."""
+        assert replica_name not in self._pending_replicas[deployment_name]
+
+        self._launching_replicas[deployment_name].pop(replica_name, None)
+        self._recovering_replicas[deployment_name].discard(replica_name)
+
+        self._running_replicas[deployment_name][replica_name] = node_id
+
+    def on_replica_recovering(self, deployment_name: str, replica_name: str):
+        """This is called whenver a deployment replica is recovering."""
+        assert replica_name not in self._pending_replicas[deployment_name]
+        assert replica_name not in self._launching_replicas[deployment_name]
+        assert replica_name not in self._running_replicas[deployment_name]
+
+        self._recovering_replicas[deployment_name].add(replica_name)
+
+    def schedule(
+        self,
+        upscales: Dict[str, List[ReplicaSchedulingRequest]],
+        downscales: Dict[str, DeploymentDownscaleRequest],
+    ):
+        """This is called for each update cycle to do batch scheduling.
```

Comment:
You are right that for spread policy, all the schedule requests can be fulfilled immediately inside the `schedule()` call. But for driver deployment policy, it might not be the case: e.g., if there are recovering replicas, scheduler won't schedule new replicas to avoid multiple replicas on the same node until future schedule() call when all replicas are recovered.